app/models/post_flag.rb

```diff
@@ -0,0 +1,51 @@
+# frozen_string_literal: true
+
+class PostFlag < ActiveRecord::Base
+  scope :enabled, -> { where(enabled: true) }
+
+  before_create :set_custom_id, unless: :system?
+  before_save :set_position
+  after_save :reset_post_action_types!
+  after_destroy :reset_post_action_types!
+
+  def used?
+    PostAction.exists?(post_action_type_id: self.id)
+  end
+
+  def self.reset_post_action_types!
```

Comment:
Something like `reset_flag_settings!` would be a nicer generic name here, we could provide a plugin registry for things like chat to hook into.